nip-39 digital identites in metadata (#201)
nostr-protocol/nips: 39.md

NIP-39
======

Digital Identities in Profile
-----------------

`draft` `optional` `author:pseudozach` `author:semisol`

## Abstract

Nostr protocol users may have other online identities such as usernames, profile pages, keypairs etc. they control and they may want to include this data in their profile metadata so clients can parse, validate and display this information.

## `identities` field in metadata content

A new optional `identities` field is introduced for `kind 0` metadata event contents in addition to name, about, picture fields as included in [NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md):
```json
{
    "id": "123",
    "pubkey": "xxx",
    ...
    "content": {
        "name": "asd",
        ...

>         "identites": [

Reviewer: "identities": [

```

An `i` tag will have two parameters, which are defined as the following:
1. `identity`: This is the identity provider name (for example `github`) and the identity  (for example `semisol`) joined together.
2. `proof`: String or object that points to the proof of owning this identity.

Clients SHOULD process any `i` tags with more than 2 values for future extensibility.  
Identity provider names SHOULD only include `a-z`, `0-9` and the characters `._-/` and MUST NOT include `:`.  
Identity names SHOULD be normalized by replacing uppercase letters with lowercase letters and if there are multiple aliases for an entity using the primary one.  
If a matching metadata event is found when trying to look up an identity, a client SHOULD check if it's the latest version, and if not if it has this claim. This is to prevent showing or checking public keys that disowned a claim.

## Claim types

### `github`

Identity: A GitHub username.  
Proof: The ID of the GitHub Gist located at `https://gist.github.com/<identity>/<proof>` with a single file with the text `Verifying that I control the following Nostr public key: <npub encoded public key>`

### `twitter`

Identity: A Twitter username.  

> Proof: The ID of the Twitter tweet located at `https://twitter.com/<identity>/status/<proof>` with the text `Verifying my account on nostr My Public Key: "<npub encoded public key>"`

Author: I specifically changed the text to the existing template that's used on nostr.directory because there are 15k+ such tweets so that can be re-used. I'll update the explanation but strongly suggest we keep the existing tweet text to allow users with existing tweets to migrate to this schema.

NIP-39
======

External Identities in Profiles
-------------------------------

`draft` `optional` `author:pseudozach` `author:Semisol`

## Abstract

Nostr protocol users may have other online identities such as usernames, profile pages, keypairs etc. they control and they may want to include this data in their profile metadata so clients can parse, validate and display this information.

## `i` tag on a metadata event

A new optional `i` tag is introduced for `kind 0` metadata event contents in addition to name, about, picture fields as included in [NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md):
```json
{

>     "id": "123",
>     "pubkey": "xxx",
>     ...

Reviewer 2: There are three options for handling stuff not touched on in the nip: omit it, show placeholders or valid data.

```suggestion
    ...
```

```suggestion
    "id": <id>,
    "pubkey": <pubKey>,
    ...
```

    "tags": [
        ["i", "github:semisol", "9721ce4ee4fceb91c9711ca2a6c9a5ab"],
        ["i", "twitter:semisol_public", "1619358434134196225"],
        ["i", "mastodon:bitcoinhackers.org/@semisol", "109775066355589974"]
        ["i", "telegram:1087295469", "nostrdirectory/770"]
    ]
}
```

An `i` tag will have two parameters, which are defined as the following:

> 1. `identity`: This is the identity provider name (for example `github`) and the identity  (for example `semisol`) joined together.

Reviewer 2: You later refer to `ìdentity` but mean `semisol` not `github:semisol` and `identity` is specified twice in this line.

```suggestion
1. `platform:identity`: This is the platform name (for example `github`) and the identity on that platform (for example `semisol`) joined together with `:`.
```

Author: I agree. @Semisol any objections to this?

2. `proof`: String or object that points to the proof of owning this identity.

Clients SHOULD process any `i` tags with more than 2 values for future extensibility.  
Identity provider names SHOULD only include `a-z`, `0-9` and the characters `._-/` and MUST NOT include `:`.  
Identity names SHOULD be normalized by replacing uppercase letters with lowercase letters and if there are multiple aliases for an entity using the primary one.  

> If a matching metadata event is found when trying to look up an identity, a client SHOULD check if it's the latest version, and if not if it has this claim. This is to prevent showing or checking public keys that disowned a claim.

Reviewer 2: Obsolete line as replaceable events like this are already defined as "the last one counts".

```suggestion
```

Author: @Semisol can you chime in please? ok to remove this line?

### `github`

Identity: A GitHub username.  
Proof: A GitHub Gist ID. This Gist should be created by `<identity>` with a single file that has the text `Verifying that I control the following Nostr public key: <npub encoded public key>`.  
This can be located at `https://gist.github.com/<identity>/<proof>`.

### `twitter`

Identity: A Twitter username.  
Proof: A Tweet ID. The tweet should be posted by `<identity>` and have the text `Verifying my account on nostr My Public Key: "<npub encoded public key>"`.  
This can be located at `https://twitter.com/<identity>/status/<proof>`.

### `mastodon`

Identity: A Mastodon instance and username in the format `<instance>/@<username>`.
Proof: A Mastodon post ID. This post should be published by `<username>@<instance>` and have the text `Verifying that I control the following Nostr public key: "<npub encoded public key>"`.  
This can be located at `https://<identity>/<proof>`.

### `telegram`

Identity: A Telegram user ID.

> Proof: A string in the format `<ref>/<id>` which points to a message published in the public channel or group with name `<ref>` and message ID `<id>`. This message should be sent by user ID `<identity>` and have the text `Verifying My Public Key: "<npub encoded public key>"`.  

Reviewer: do we really need the string `Verifying that I control the following Nostr public key:` everytime? isn't nostr already implied by just using `<npub encoded public key>`? 
if we do need it, then it would be better if it's always the same and here it's different in this telegram block.

Author: it would be confusing for people on other social sites to see just a pubkey with no context. Having a grammatically correct statement is a good idea in my view. it's sort of a written contract/claim.

Reviewer 2: Then still it makes sense to use the same sentence. I assume you accidentally had variations there.

Author: @Giszmo no, tweet and telegram text is different on purpose because of 16k+ tweets+tg posts that already have the text from nostr.directory. I think we should keep it to avoid having people do the same thing again. Or if we must keep the text the same then maybe we can put a note so clients can also accept variations on the wording of the statement.

Reviewer 3: just and idea, but for github and similar sites a SSH pubkey could be used to more directly proof ownership. You could use openssl or smth to sign a message with the SSH identity that is visible on github (the pubkey can be found at a specific url. Similar for sr.ht etc)

This would then use the encoded signature as the third element of the tag array. Pubkeys are usually Ed25519 or RSA signatures, so clients would have to implement cryptography instead of using https+gist to verify validity.

This could be 'github+ssh:account', say.

Reviewer 3: Just came back to say this comment above ^ is nonsensical; you would still need to get the file containing the pubkey and the. *additionally* do crypto. I'll leave this note here in case anyone else has this same brilliant idea haha.

gist + well known string = KISS
